keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/AcceptanceRateTracker.java

```diff
@@ -0,0 +1,34 @@
+package io.improbable.keanu.algorithms.mcmc.proposal;
+
+import com.google.common.collect.Maps;
+import io.improbable.keanu.vertices.Vertex;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+public class AcceptanceRateTracker implements ProposalListener {
+
+    private Map<Set<Vertex>, Integer> numApplied = Maps.newHashMap();
+    private Map<Set<Vertex>, Integer> numRejected = Maps.newHashMap();
+
+    @Override
+    public void onProposalApplied(Proposal proposal) {
+        Set<Vertex> key = proposal.getVerticesWithProposal();
+        log.trace(String.format(" Applied proposal: %.4f", getAcceptanceRate(key)));
```

Comment:
This getAcceptanceRate(key) will get called even if log.trace doesn't log anything, which means there's a performance cost to logging that may never be used and you pay the cost even when it's not used. This is one reason I don't like liberal logging all over the place. The other reason is YAGNI but that's more debatable. There's probably a clever way around this issue but we should be mindful of this sort of logging.

Reply:
Ah, that's an interesting point. 

I would definitely debate "YAGNI" because I already did need it! However - it'll always be possible for the calling code to do the logging (especially with streaming samples). So I 
will remove it from this class.